prov/efa/src/efa_cq.c

```diff
@@ -0,0 +1,535 @@
+			 fi_strerror(-ret));
+		efa_base_ep_write_eq_error(base_ep, -ret,
+					   FI_EFA_ERR_WRITE_SHM_CQ_ENTRY);
+	}
+
+	efa_cntr_report_rx_completion(&base_ep->util_ep, flags);
+}
+
+/**
+ * @brief poll rdma-core cq and process the cq entry
+ *
+ * @param[in]	cqe_to_process	Max number of cq entry to poll and process. A negative number means to poll until cq empty
+ * @param[in]   cq      efa_cq
+ */
+void efa_cq_poll_ibv_cq(ssize_t cqe_to_process, struct util_cq *util_cq)
+{
+	bool should_end_poll = false;
+	struct efa_base_ep *base_ep;
+	struct efa_cq *cq;
+	struct efa_domain *efa_domain;
+	struct fi_cq_tagged_entry cq_entry = {0};
+	struct fi_cq_err_entry err_entry;
+	ssize_t err = 0;
+	size_t num_cqe = 0; /* Count of read entries */
+	int prov_errno, opcode;
+
+	/* Initialize an empty ibv_poll_cq_attr struct for ibv_start_poll.
+	 * EFA expects .comp_mask = 0, or otherwise returns EINVAL.
+	 */
+	struct ibv_poll_cq_attr poll_cq_attr = {.comp_mask = 0};
+
+	cq = container_of(util_cq, struct efa_cq, util_cq);
+	efa_domain = container_of(cq->util_cq.domain, struct efa_domain, util_domain);
+
+	/* Call ibv_start_poll only once */
+	err = ibv_start_poll(cq->ibv_cq.ibv_cq_ex, &poll_cq_attr);
+	should_end_poll = !err;
+
+	while (!err) {
+		base_ep = efa_domain->qp_table[ibv_wc_read_qp_num(cq->ibv_cq.ibv_cq_ex) & efa_domain->qp_table_sz_m1]->base_ep;
+		opcode = ibv_wc_read_opcode(cq->ibv_cq.ibv_cq_ex);
+		if (cq->ibv_cq.ibv_cq_ex->status) {
+			prov_errno = ibv_wc_read_vendor_err(cq->ibv_cq.ibv_cq_ex);
+			switch (opcode) {
+			case IBV_WC_SEND: /* fall through */
+			case IBV_WC_RDMA_WRITE: /* fall through */
+			case IBV_WC_RDMA_READ:
+				efa_cq_handle_error(base_ep, cq->ibv_cq.ibv_cq_ex,
+						    base_ep->util_ep.tx_cq,
+						    to_fi_errno(prov_errno),
+						    prov_errno);
+				break;
+			case IBV_WC_RECV: /* fall through */
+			case IBV_WC_RECV_RDMA_WITH_IMM:
+				if (efa_cq_wc_is_unsolicited(cq->ibv_cq.ibv_cq_ex)) {
+					EFA_WARN(FI_LOG_CQ,
+						 "Receive error %s (%d) for unsolicited write recv",
+						 efa_strerror(prov_errno), prov_errno);
+					efa_base_ep_write_eq_error(
+						base_ep,
+						to_fi_errno(prov_errno),
+						prov_errno);
+					break;
+				}
+				efa_cq_handle_error(base_ep, cq->ibv_cq.ibv_cq_ex,
+						    base_ep->util_ep.rx_cq,
+						    to_fi_errno(prov_errno),
+						    prov_errno);
+				break;
+			default:
+				EFA_WARN(FI_LOG_EP_CTRL, "Unhandled op code %d\n", opcode);
+				assert(0 && "Unhandled op code");
+			}
+			break;
+		}
+
+		cq->read_entry(cq->ibv_cq.ibv_cq_ex, &cq_entry);
+
+		switch (opcode) {
+		case IBV_WC_SEND: /* fall through */
+		case IBV_WC_RDMA_WRITE: /* fall through */
+		case IBV_WC_RDMA_READ:
+			efa_cq_handle_send_completion(
+				base_ep, cq->ibv_cq.ibv_cq_ex, &cq_entry);
+			break;
+		case IBV_WC_RECV:
+			efa_cq_handle_recv_completion(
+				base_ep, cq->ibv_cq.ibv_cq_ex, &cq_entry);
+			break;
+		case IBV_WC_RECV_RDMA_WITH_IMM:
+			efa_cq_proc_ibv_recv_rdma_with_imm_completion(
```

Comment:
I think you can use a general function to cover all
`efa_cq_handle_send_completion`, `efa_cq_handle_send_completion`, and `efa_cq_proc_ibv_recv_rdma_with_imm_completion `, if you can use a simple logic to tell whether to write completion based on wr_id.

Reply:
Think about it again, probably worth a special function form recv rdma with imm, as you will always write a cq entry for it.